JBPM-8118: Containment is never allowed when multiple selection.

romartin opened the pull request:
Hey @hasys @tiagodolphine 

The goals for this commit is to fix several issues at once: 
* [JBPM-8118](https://issues.jboss.org/browse/JBPM-8118)
* [JBPM-8182](https://issues.jboss.org/browse/JBPM-8182)
* [JBPM-8105](https://issues.jboss.org/browse/JBPM-8105)
* [JBPM-8158](https://issues.jboss.org/browse/JBPM-8158)
* [JBPM-8235](https://issues.jboss.org/browse/JBPM-8235)
* ~~[JBPM-7099](https://issues.jboss.org/browse/JBPM-7099)~~ _Finally not fixed here, see [comment](https://github.com/kiegroup/kie-wb-common/pull/2451#issuecomment-462953019)_


Some notes about what this commit does:
- Refactor due to new lienzo's `WiresLayerIndex` stuff (see https://github.com/kiegroup/lienzo-core/pull/88)
- Fix undo/redo issues due to bad composite command undo direction (RequestCommandManager)
- Refactor `toString` methods for all canvas commands (logging was terrible!)
- Improve client "dev" commands logging
- Refactored commands & components related to Control Points - @tiagodolphine I did this because of different reasons: 
-- To delete the `isCanvasCommandFirst` stuff in `AbstractCanvasGraphCommand`. It was difficult to follow on top of abstract commands, which are not easy as well... so I tried some different approach
-- To avoid [this change](https://github.com/kiegroup/kie-wb-common/pull/2035/files#diff-914d5b1cceaa40ab6aa47bbc1c760e70L219), as it brakes other stuff, like containment or docking.  The assumption is that the undo direction when using the `RequestCommandManager` is `reverse`
-- Also now the Stunner's `IControlPointsAcceptor` implementation just fires a single command when user adds a new CP and also drags it on the same mouse operation. It was firing two different commands before, which must be reverted in `forward` direction, and well it's fine.. but it was getting difficult to integrate with the `RequestCommandManager`, which defaults to a `reverse` direction
-- Finally noticed that the model for CPs was a bit confusing.. the CPs for an instance where handled by a `List`, but  also each CP was declaring an `index` field. Also some additional computation was being done (and tested) to manage this. Not sure if there was some reason for it, probably I could be missing something, but tried to simplify it, by removing the `index` from the CP class and refactoring the `List` into an `array`, so order is implicit.. does that makes sense?

Soo please as for this PR you should check containment, docking (single & multiple nodes) and also play around with control points.

Assembly:
https://github.com/kiegroup/lienzo-core/pull/88
https://github.com/kiegroup/lienzo-tests/pull/63

Thanks!!

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

looks great, so far I found those issues (but some of them are reproducible on master):

1. Still not possible to move full process, like Start→End inside of Sub-process. Probably, can be handled in separate issue: [JBPM-7629](https://issues.jboss.org/browse/JBPM-7629)
2. Not possible to move only part of selected nodes outside of Swimlane. For example, if you have Start→End couple, it is not possible to place Start on Process level and End inside of Swimlane.
3. Append cannot be applied near to the right border of a swimlane. 
4. Sometimes sequence flow remains under the container, see image below I just moved container around. During move all sequence flows going under swimlane, but after drag is finished they moved back to the top of the container. But now some sequence flows remains under container. 
**pic 1**
![picture 1](https://user-images.githubusercontent.com/1477262/52429051-9e2a9d80-2b03-11e9-9618-65adf019a35f.png)
5. Model Start→Task→End inside of the Embedded sub-process, add Swimlane to the process and move entire Embedded sub-process inside of Swimlane. See pic 2 for actual result. Move Embedded sub-process back to the process level. See pic 3 for the result (all sequence flows will be under the sub-process, but pic. 3 can be reproduced on master as well.)
**pic 2**
![picture 2](https://user-images.githubusercontent.com/1477262/52429137-c0242000-2b03-11e9-8f1b-92dfa5173d68.png)
**pic 3**
![picture 3](https://user-images.githubusercontent.com/1477262/52429148-c31f1080-2b03-11e9-88fc-3c81fddbb16d.png)
6. Add Container, inside of the container model simple Start→End process, remove Sequence flow and try to move the container. 
  * Actual Result: NullPointer, you need to reopen process to continue your work. (I am able to reproduce it on master as well)
  *  Uncaught exception: Exception caught: (TypeError) : Cannot read property 'Qc' of null Caused by: (TypeError) : Cannot read property 'Qc' of null
* Model process pic 4, try to delete Task, exception will be shown. (I am able to reproduce it on master branch as well)
![image](https://user-images.githubusercontent.com/1477262/52429212-e2b63900-2b03-11e9-9765-4a072cb65f60.png)

So, there are some issues, but most of them are reproducible on master too, so we can move them to separate Jira issues. For now regressions should be fixed: 4 (little bit tricky to reproduce) and 5 issue shown on pic 2. The second part of the 5-th issue is reproducible on master.

Also, we should at least consider fix 1-3 issues in this PR as well.

I will perform more tests with undo/redo tomorrow.

hasys commented:
@romartin,

another question: looks like you fixed not only JBPM-8118, but also other issues like JBPM-8182. Can I ask you to provide full list of fixed issues to simplify testing?

romartin commented:
Hey @tiagodolphine @hasys 

Great, working on your feedback now, thanks for the review!! :+1: 

@hasys hehe good point, I was missing to review "what" exactly this PR fixes, as it was not only JBPM-8118. Also good catch on the overall testing :)

Soo my goal for this PR is trying to fix as much as issues as possible at once. We're close to the release and I would like to have this merged ASAP to get some time to other to tests all changes, so I'm trying to do my best to fix as much issues as possible here but also considering not delivering this last minute. 

Saying this, @hasys let's try to fix the issues here!! (feeling brave for a while :1st_place_medal: ) 

Here are my answers:
* About 1 - agree, that issue is not in scope for this PR, my plan is to fix it next week, but I would prefer doing this task in an isolated way, it's quite core. So let's forget this issue for now, I moved [JBPM-7629](https://issues.jboss.org/browse/JBPM-7629)  on top of backlog, to keep track of it

* About 2 - That's how it works, I mean it's intentionally coded like this - the bounding box computed by all selected shapes must fit in the "target" container, in order to allow multiple containment. I don't say it's the right thing to do, but let's also put this of out scope of this PR, feel free to report a ticket for it

* About 3 - Well this is a bit tricky and also buggy, I think it already exists some reported ticket for this. The thing is that if you try to append a new node close  to the border of a container, and the new node does "not fit", it looks if it's possible to append the new node on a "new line/row", this way, by doing the increasing the height of the container, and clicking again, the new node should appear on below... Anyway this code is also buggy, and not easy to follow, so also out of scope of this PR

* About 4 and 5 -  both are related to the connector's shape visibility when using containers, right? I  think we can address these issues in this PR also, it's supposed not to be hard, and I also think are the same issues  as [JBPM-7099](https://issues.jboss.org/browse/JBPM-7099) describes, right? Anyway, good catch, lemme work on this :+1: 

* About 6 - Agree - lemme work on this and also try to fix it

So summarizing, I'll work on the **4**, **5** and **6**, and also I found this **bug**, which I'll fix:
1. Append a timer event boundary into a user task
2. Move the user task (so also the timer event) into a container (lane)
3. Undo -> produces an error

PS: Also updated this ticket's description in order to describe the tickets which it fixes.

Thanks guys, great job!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/command/RequestCommandManager.java:
```diff
@@ -206,21 +202,16 @@ private void start() {
-                LOGGER.log(Level.FINEST,
-                           "Adding commands for current request into registry [size=" + commands.size() + "]");
                 getRegistry().register(new CompositeCommand.Builder<AbstractCanvasHandler, CanvasViolation>()
+                                               .forward() // PLEASE DO NOT CHANGE THIS!
```
hehe good catch :+1:  Updating the comment, let's see if it helps a bit more...

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/wires/WiresCanvasView.java:
```diff
@@ -93,6 +93,7 @@ public LienzoCanvasView addChild(final ShapeView<?> parent,
         final WiresShape childShape = (WiresShape) child;
         layer.addChild(parentShape,
                        childShape);
+        childShape.shapeMoved();
```
yeahh absolutely agree man, and even more for this `shapeMoved` method!! I see it everywhere... anyway it's a bit tricky this change, and tbh I would prefer keeping this as an improvement, so better assuming now some more computation time than a bug. IMO the extra computation time spent on multiple calls to this method can be assumed for now as we don't expect hard diagram hierarchies, so parent/children iterations are gonna to be still quite fast...

romartin commented:
Hey @hasys @tiagodolphine 

Let you know that the error I reported on the comment above when having a boundary timer event, is the same one as @LuboTerifaj  already reported in [JBPM-8218](https://issues.jboss.org/browse/JBPM-8218), and it's also a duplicate for [JBPM-8214](https://issues.jboss.org/browse/JBPM-8214).

So once [this PR](https://github.com/kiegroup/kie-wb-common/pull/2459) from @xiezhang7 gets merged, it will be fixed. I've also tested locally, so let's consider removing this issue out of scope from this PR. 

Thanks!

romartin commented:
Hey @hasys @tiagodolphine 

I've been working on the fixes from your comments, both on the code improvements and also on the issues reported, also rebased from master branch.

Summary:
**Issues 4 & 5** 
Due to the implementation on wires side, actually connectors are not part of any shape's container, and added into the layer directly, this does not allows to move connector up to the right parent, in case of multiple hierarchy, anyway assuming this, I've fixed it by moving connector on top, no matter the parent. l

**Issue 6(A)**
> Add Container, inside of the container model simple Start→End process, remove Sequence flow and try to move the container
Fixed

**Issue 6(B)**
> Model process pic 4, try to delete Task, exception will be shown. (I am able to reproduce it on master branch as well)
I've been analyzing it, and the root issue for this is on the rules/command evaluations, it can be addressed as with [JBPM-7629](https://issues.jboss.org/browse/JBPM-7629) because it's the same root problem. So part of this issue I'll better fix in the JBPM-7629.
Anyway I've fixed the "canvas" side, at although when trying to remove the task some rule validation error appears, it does not produce any runtime exception and the user is able to, at least, continue working and do not loose any changes.


Also notice the bug reported by me, about a boundary timer event, will be fixed once merging another PR - see comment above.

Thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

I noticed great performance improvements on my laptop. Will play more with it in different browsers/laptops, but everything looks very smooth now, great job!

About issues, so far I found two quite major regressions:

### First issue:
1. Create a process, append Task to initial Start event
![image](https://user-images.githubusercontent.com/1477262/52580344-6e8fd400-2e28-11e9-8699-342b28866ce5.png)
2. Click undo and than redo buttons
![image](https://user-images.githubusercontent.com/1477262/52580370-7780a580-2e28-11e9-8502-ddda3c0ca1be.png)
3. Click undo and than redo several times
  * **Results:**
Sequence flow from Start node side will be attached to different points after every undo/redo round. After several iterations of undo/redo you will see an Error and diagram will be corrupted.
![image](https://user-images.githubusercontent.com/1477262/52580377-7d768680-2e28-11e9-8b26-ecdb0a952d46.png)
![image](https://user-images.githubusercontent.com/1477262/52580394-85362b00-2e28-11e9-8e39-32b5704f6b8b.png)
![image](https://user-images.githubusercontent.com/1477262/52580405-89fadf00-2e28-11e9-909d-0bc989d82cf1.png)
![image](https://user-images.githubusercontent.com/1477262/52580424-941cdd80-2e28-11e9-854c-a2c1c4497921.png)

### Second issue:
1. Create a process, add Embedded sub-process
2. Drag and drop from the pallet Start, Task, End to the sub-proccess
3. Connect nodes using quick menu
4. Move sub-process
  * **Result**
Sequence flow from Start to Task will preserve it's position and won't move with sub-process.
![image](https://user-images.githubusercontent.com/1477262/52580185-1c4eb300-2e28-11e9-8bc3-513ae6b86f27.png)
5. Try to connect Start and Task one more time, try to move sub-process .
  * **Result**: Top bendpoints will be used instead of center one.
![image](https://user-images.githubusercontent.com/1477262/52580229-38eaeb00-2e28-11e9-87e7-8c2986d9706b.png)
6. Try to move sub-process one more time
  * **Result**: Start events bendpoint will be attached to the process
![image](https://user-images.githubusercontent.com/1477262/52580252-443e1680-2e28-11e9-9c4c-81758fd83a7a.png)
7. Try to connect disconnected sequence flow with Start again
  * **Result**: Error will be shown, process will disappear:
![image](https://user-images.githubusercontent.com/1477262/52580115-fb865d80-2e27-11e9-9a60-62fb59eaa743.png)

romartin commented:
Hey @hasys 
Thanks for testing, I think both bugs you reported can be due to same reason.. and it also looks it's some "last" time fix I commited, so trying to fix it and do some tests to not broke it again.
Thanks!!!

romartin commented:
Hey @hasys @tiagodolphine 

Just fixed the issues that @hasys reported about magnets. This fix has been done on lienzo side, so updated https://github.com/kiegroup/lienzo-core/pull/88 and https://github.com/kiegroup/lienzo-tests/pull/63 by adding the fix and rebased both PRs.

Also introduced a new commit here (`kie-wb-common`) that fixes [JBPM-8158](https://issues.jboss.org/browse/JBPM-8158)

So ready for testing again!! :+1: 

Thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented:
war files didn't generated, looks like random fail, or build were executed before lienzo-core were updated.

Jenkins execute full downstream build

hasys commented:
Hi @romartin,

not sure what is going on, but Fulldowstream failed for the second time in the row with the same issue on Showcase project. I compiled Standalone locally and testing it now, but still we should to find out why build is failing and make a quick check with full build.

tiagodolphine commented:
Strange because the project build was ok, only the full downstream is red.

hasys commented:
Hi @romartin, I tested Standalone project, you can find results below. But, I still need the full build to test one of the fixed issue and to test integration/different browsers.

### Tests by fixed jiras:
[JBPM-8118](https://issues.jboss.org/browse/JBPM-8118) - works good
[JBPM-8182](https://issues.jboss.org/browse/JBPM-8182) - works good
[JBPM-8105](https://issues.jboss.org/browse/JBPM-8105) - works good, but I need to test it with war as well
[JBPM-7099](https://issues.jboss.org/browse/JBPM-7099) - still reproducible, see picture below:
![disappearedsequenceflows](https://user-images.githubusercontent.com/1477262/52644584-6050bf00-2edf-11e9-89d7-e02c288c4119.png)
[JBPM-8158](https://issues.jboss.org/browse/JBPM-8158) - works good

### Found regressions:
1. When I undo container resize, resize control point position is not Undo:
  ** Steps to reproduce: add swimlane, resize, undo two times, one time redo
![image](https://user-images.githubusercontent.com/1477262/52642977-5aa5aa00-2edc-11e9-978a-c8d97ef1ca2c.png)
2. When I moving node out of the subprocess sequence flow changes it's color to red. But when you return node to the correct position red color is not reset back to black. You need to select different node to update sequence flow state. (it is regression, it was correct before)
![image](https://user-images.githubusercontent.com/1477262/52643129-a5272680-2edc-11e9-9aeb-be0e153ac8bb.png)
3. Any move of the container activates all inner sequence flows and sequence flows connected to this subprocess:
![image](https://user-images.githubusercontent.com/1477262/52643554-955c1200-2edd-11e9-920d-d66cf8c0ec65.png)

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
@hasys now the wars were generated ok. 
I think it was a random error, causing jvm crash before.

romartin commented:
thanks @hasys @tiagodolphine again, working on the fixes :+1:

romartin commented:
Hey @hasys , about [JBPM-7099](https://issues.jboss.org/browse/JBPM-7099) you're right, it's not fixed, 

I just fixed the connectors appear on top on the "drag end". There exist a "limitation" now for moving connectors up to the right "parents", and it's also not working while dragging, see [comment here](https://issues.jboss.org/browse/JBPM-7099?focusedCommentId=13694622&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13694622), so for now, I just fixed by moving connectors on top once drag ends. So will keep JBPM-7099 open.

Working on the other stuff now, great job, thanks a lot!

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @tiagodolphine @hasys 

Thanks for the review!! Here is the summary again:
* Removing [JBPM-7099](https://issues.jboss.org/browse/JBPM-7099) out of scope from this PR, although this PR fixes the connectors by making them visible when multiple parents are nested
* About regression 1 -> is not a regression, I just created [JBPM-8235](https://issues.jboss.org/browse/JBPM-8235) for this. Anyway I fixed the issues found! @tiagodolphine Finally refactored most of the stuff from `ControlPointsControl` into a new command yeah, so making it more isolated and also allows to fix the issues found easily, hope it looks good for you :)
* About regression 2 -> Fixed the sequence flow, it's not back to normal color once dragging back to the subprocess. Also notice the end event shape on the screenshot was blue, this was not expected as well, so fixed it too
* About regression 3 -> Fixed too!

There is lot of work here guys, IMO now it's time to postpone small improvements and focus on merging this, it provides lot of fixes that I don't want to introduce last day, if possible. So if errors appear and make sense to fix them now, great, I'll be happy to do it, but please consider the urgency on those decisions.

Anyway great job on the review, thanks!! :+1:

hasys commented:
Hi @romartin,

works very solid, great job! Tested listed issues in Chrome/Firefox/IE11.

But I found some issues:

1. Related to issue 6 from this comment: https://github.com/kiegroup/kie-wb-common/pull/2451#issuecomment-461517594

It is not possible to remove task from process as you can see below, error will be shown. It's better than before, because it doesn't throw exception, but still it should be possible to remove the task.
![image](https://user-images.githubusercontent.com/1477262/52728760-2c959800-2fb8-11e9-9458-678f81fd679b.png)
![image](https://user-images.githubusercontent.com/1477262/52728854-551d9200-2fb8-11e9-92aa-fe39b1e51eca.png)

2. I see an exception `Uncaught exception: Exception caught: (TypeError) : this.p is null Caused by: (TypeError) : this.p is null` when I trying to put any _Service Task_ very close to a container border. If you place Service Task to large free space no exceptions shown and task added. It broke nothing, but annoying exception dialog with exception message is shown.
![image](https://user-images.githubusercontent.com/1477262/52729315-37046180-2fb9-11e9-81b4-10412aafd378.png)

3. Probably unrelated to this PR: if you have sub-process with content (for an instance Embedded) and try to morph it to another sub-process type (for an instance Event) all content will be lost and sequence flows will be dislocated and it is not possible to Undo this action. See pictures:
**Before**
![image](https://user-images.githubusercontent.com/1477262/52732142-bdbc3d00-2fbf-11e9-9a38-4d635f18c2ac.png)
**After**
![image](https://user-images.githubusercontent.com/1477262/52732161-c9a7ff00-2fbf-11e9-9fa8-86ab6869fdba.png)
P.S. I found a possibility to undo, you can remove sub-process and Undo it, everything looks fine after that:
![image](https://user-images.githubusercontent.com/1477262/52732200-ec3a1800-2fbf-11e9-9123-18275147c8b7.png)
P.P.S If you try to Undo once before removing sub-process -> sub-process type will be restored, but nodes not, if you remove sub-process now and Undo -> initial sub-process will be restored with all nodes.

4. As you mentioned on IRC there is a bug in CM, it is not possible to drag & drop tasks.

romartin commented:
Hey @hasys 

We're working hard on this PR trying to make the release stable, it's a great job, so thanks for the collaboration and review again!! :+1: 

Rebased and added new commits for the issues [above](https://github.com/kiegroup/kie-wb-common/pull/2451#issuecomment-463298191). Summary:

**Issue 1**
Good point. I've decided to do as we did for cardinality rules for nodes in general, but now also for edges (connectors). So if some cardinality rule now fails, instead of blocking the UI in order to fix it, the violations are both type `WARNING`, which means they appear when validating/saving the process, but they allow to continue the work on the canvas to fix them. So good catch, it was missing to do the same for the connectors.
For example, looking at the _picture 1_ you shared, now it's possible to remove the sequence flow from the blue timer event and keep working with it. But as there exist [this rule](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/SequenceFlow.java#L63) for this event, in case it has no incoming connections or has more than one, once validating the process it will show a warn to the user:

![screenshot from 2019-02-14 02-26-54](https://user-images.githubusercontent.com/4602417/52757860-786f2e00-3006-11e9-88dd-da181ec393cd.png)

Again, same solution with did time ago with, for example, end events in a process, it allows to save a process with no end event, but it warns about it. Same here for sequence flows.

**Issue2**
Yeah you're right... it's strange and buggy, so TBH I applied a quick win, but IMO looks good. It's about same as above, now the OutOfBoundErrors are `WARNING` types instead of `ERROR`. So fixed this way :+1: 

**Issue3**
Good catch! But well, it's not related... let's keep this issue for later work, probably we can just report a new ticket. WDYT?

Thanks!
Roger

romartin commented:
Jenkins execute full downstream build

romartin commented:
@hasys the regular J build failed due those tests failures that @wmedvede @tiagodolphine @xiezhang7 are looking at, not much I can do here.. so let's try to build again

romartin commented:
Jenkins please retest this

wmedvede commented:
@romartin @hasys 
The artifacts were properly generated in the last try.

This was the as usually failing tests:
Test Result (11 failures / ±0)
org.kie.wb.selenium.ui.LoadAllPerspectivesIntegrationTest.allPerspectivesCanBeLoaded
org.jbpm.designer.web.server.TransformerServletTest.testScalePDFImage[0]
org.jbpm.designer.web.server.TransformerServletTest.testSharePdf[0]
org.jbpm.designer.web.server.TransformerServletTest.testSharePng[0]
org.jbpm.designer.web.server.TransformerServletTest.testTransformToPdf[0]
org.jbpm.designer.web.server.TransformerServletTest.testTransformToPng[0]
org.jbpm.designer.web.server.TransformerServletTest.testScalePDFImage[1]
org.jbpm.designer.web.server.TransformerServletTest.testSharePdf[1]
org.jbpm.designer.web.server.TransformerServletTest.testSharePng[1]
org.jbpm.designer.web.server.TransformerServletTest.testTransformToPdf[1]
org.jbpm.designer.web.server.TransformerServletTest.testTransformToPng[1]

hasys commented:
Hi @romartin,

1. and 2. looks good as temporary fix, but I think it will be better to return it to Error state and fix existing issues as well in the future. Let's report it to the Jira if you agree.

Also, I filed unrelated issue: [JBPM-8246](https://issues.jboss.org/browse/JBPM-8246).

tiagodolphine commented:
@romartin just to say, I tested and it worked pretty good, the perfomance mainly with control points is much butter and the resize command was a great code improvements.
Thanks.
